go/mysql/binlog_event_compression.go

```diff
@@ -0,0 +1,254 @@
+/*
+Copyright 2023 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mysql
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/klauspost/compress/zstd"
+
+	"vitess.io/vitess/go/vt/log"
+	vtrpcpb "vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/vterrors"
+)
+
+// This file contains code related to handling compression related
+// events. More specifically today, compressed transaction payloads:
+// See: https://dev.mysql.com/doc/refman/en/binary-log-transaction-compression.html
+
+// Transaction Payload wire protocol fields:
+// https://dev.mysql.com/doc/dev/mysql-server/latest/classbinary__log_1_1codecs_1_1binary_1_1Transaction__payload.html
+const (
+	payloadHeaderEndMark = iota
+	payloadSizeField
+	payloadCompressionTypeField
+	payloadUncompressedSizeField
+)
+
+// Compression algorithms that are supported (only zstd today
+// in MySQL 8.0):
+// https://dev.mysql.com/doc/refman/8.0/en/binary-log-transaction-compression.html
+const (
+	transactionPayloadCompressionZstd = 0
+	transactionPayloadCompressionNone = 255
+)
+
+var transactionPayloadCompressionTypes = map[uint64]string{
+	transactionPayloadCompressionZstd: "ZSTD",
+	transactionPayloadCompressionNone: "NONE",
+}
+
+// Create a reader that caches decompressors.
+var zstdDecoder, _ = zstd.NewReader(nil, zstd.WithDecoderConcurrency(0))
```

Comment:
Is this concurrency safe? Would we ever have multiple decodings happening in parallel?

Reply:
Ah wait, looks like https://pkg.go.dev/github.com/klauspost/compress/zstd#NewReader answers this as we're using `DecodeAll`.

Would it be useful at some point to do streaming mode? Since from what I understand a single encrypted entry can contain multiple events, we would already emit events while decompressing so we don't have to decompress everything into memory?

Reply:
I was using the streaming mode at first and it's SUPER slow. It would make sense to potentially switch to that based on the uncompressed payload size.... maybe we switch to streaming if it's > 128MiB or something?

Reply:
@mattlord Hmm, interesting. How slow is super slow? Seems like maybe something we should dig into at some point but it doesn't have to hold up this PR I think.

Reply:
Not THAT slow, but slow enough that it caused failures in the OnlineDDL VRepl stress tests — which have a bunch of threads executing small transactions against the DB while the workflow is running — as the vstream could not keep up enough to perform the final cutover.

Reply:
@mattlord That reminds me. I think we should update to the latest https://github.com/klauspost/compress as well here, looks like there recently were quite a few `ztsd` improvements.